src/cli/commands/list.js

```diff
@@ -180,21 +181,33 @@ export function filterTree(tree: Tree, filters: Array<string>): boolean {
   return notDim && (found || hasChildren);
 }
 
+const getDevDeps = (manifest): Array<string> => Object.keys(manifest.devDependencies).map(key => `${key}@${manifest.devDependencies[key]}`);
+
 export async function run(config: Config, reporter: Reporter, flags: Object, args: Array<string>): Promise<void> {
   const lockfile = await Lockfile.fromDirectory(config.lockfileFolder, reporter);
   const install = new Install(flags, config, reporter, lockfile);
-  const {requests: depRequests, patterns, workspaceLayout} = await install.fetchRequestFromCwd();
+
+
+  const {requests: depRequests, patterns, manifest, workspaceLayout} = await install.fetchRequestFromCwd();
   await install.resolver.init(depRequests, {
     isFlat: install.flags.flat,
     isFrozen: install.flags.frozenLockfile,
     workspaceLayout,
   });
 
+  let pkgPatterns = [];
+  if (isProduction(process.env)) {
+    const devDeps = getDevDeps(manifest);
+    pkgPatterns = patterns.filter(pattern => !devDeps.includes(pattern));
```

Comment:
Hmm ... 

So, let's say `a` is our dependencies and `b` is our dev dependencies.

For production, we're looking for the subset, `c`, which is found by `a` - `b`

In both cases where `a` > `b` and `b` > `a` you would have to create a set of the largest ( or smallest depending on where you want to take the linear hit ) and then iterate over the other collection to find the subset of `a` - `b`.

I think converting `devDeps` to a hash will get us to O(n) + m, which is pretty _okay_ :)